storage/bootstrap/bootstrapper/commitlog/source.go

```diff
@@ -166,93 +257,182 @@ func (s *commitLogSource) Read(
 					lastWriteAt: dp.Timestamp,
 					enc:         enc,
 				})
-				unmergedSeries.encoders[blockStart] = unmergedBlock
+				unmergedSeries.encoders[blockStart.UnixNano()] = unmergedBlock
 			}
 		}
 		if err != nil {
-			errs++
+			workerErrs[workerNum]++
 		}
 	}
-	if errs > 0 {
-		s.log.Errorf("error bootstrapping from commit log: %d block encode errors", errs)
+	wg.Done()
+}
+
+func (s *commitLogSource) shouldEncodeSeries(
+	namespace ts.ID,
+	shardsTimeRanges result.ShardTimeRanges,
+	blockSize time.Duration,
+	series commitlog.Series,
+	dp ts.Datapoint,
+) bool {
+	// Check if the series belongs to current namespace being bootstrapped
+	if !namespace.Equal(series.Namespace) {
+		return false
+	}
+
+	// Check if the shard is one of the shards we're trying to bootstrap
+	ranges, ok := shardsTimeRanges[series.Shard]
+	if !ok {
+		return false
+	}
+
+	// Check if the block corresponds to the time-range that we're trying to bootstrap
+	blockStart := dp.Timestamp.Truncate(blockSize)
+	blockEnd := blockStart.Add(blockSize)
+	blockRange := xtime.Range{
+		Start: blockStart,
+		End:   blockEnd,
+	}
+
+	return ranges.Overlaps(blockRange)
+}
+
+func (s *commitLogSource) printM3TSZEncodingOutcome(workerErrs []int, iter commitlog.Iterator) {
+	errSum := 0
+	for _, numErrs := range workerErrs {
+		errSum += numErrs
+	}
+	if errSum > 0 {
+		s.log.Errorf("error bootstrapping from commit log: %d block encode errors", errSum)
 	}
 	if err := iter.Err(); err != nil {
 		s.log.Errorf("error reading commit log: %v", err)
 	}
+}
 
-	errs = 0
-	emptyErrs := 0
+func (s *commitLogSource) mergeShards(
+	numShards int,
+	bopts result.Options,
+	blopts block.Options,
+	encoderPool encoding.EncoderPool,
+	unmerged []map[ts.Hash]encodersByTime,
+) result.BootstrapResult {
+	shardErrs := make([]int, numShards)
+	shardEmptyErrs := make([]int, numShards)
 	bootstrapResult := result.NewBootstrapResult()
 	blocksPool := bopts.DatabaseBlockOptions().DatabaseBlockPool()
 	multiReaderIteratorPool := blopts.MultiReaderIteratorPool()
+	// Controls how many shards can be merged in parallel
+	mergeSemaphore := make(chan bool, s.opts.MergeShardsConcurrency())
+	bootstrapResultLock := sync.Mutex{}
+	wg := &sync.WaitGroup{}
```

Comment:
Don't necessarily need to make this a pointer type, it'll be captured by ref anyhow by the goroutine with just `var wg sync.WaitGroup`

Reply:
Also would replace all these definitions with a var block:

```
var (
	shardErrs = make([]int, numShards)
	shardEmptyErrs = make([]int, numShards)
	bootstrapResult = result.NewBootstrapResult()
	blocksPool = bopts.DatabaseBlockOptions().DatabaseBlockPool()
	multiReaderIteratorPool = blopts.MultiReaderIteratorPool()
	mergeSemaphore = make(chan bool, s.opts.MergeShardsConcurrency())
	bootstrapResultLock  sync.Mutex{}
	wg sync.WaitGroup{}
)
```